utf8.c

```diff
@@ -1685,8 +1685,55 @@ Perl__utf8n_to_uvchr_msgs_helper(const U8 *s,
             *msgs = NULL;
         }
 
+
+/* Returns 0 if no message need to be generated for this problem even if
+ * everything else says to?  Otherwise returns the warning category to use for
+ * the message.
+ *
+ * No message is generated if the UTF8_CHECK_ONLY flag has been set by the
+ * caller.  Otherwise, a message is generated if either:
+ *  1)  the caller has furnished a structure into which messages should be
+ *      returned to it (so it itself can decide what to do); or
+ *  2)  warnings are enabled for either of the the category parameters to the
+ *      macro.  Warnings for the passed-in category are
+ *
+ * The 'warning' parameter is the higher priority warning category to check.
+ * The macro calls ckWARN_d(warning), so warnings for it are considered to be
+ * on by default.
+ *
+ * The second, lower priority category is optional.  To specify not to use one,
+ * call the macro like:   NEED_MESSAGE(WARN_FOO, , 0)'
+ * Otherwise like:        NEED_MESSAGE(WARN_FOO, ckWARN_d, WARN_BAR)
+ *
+ * The second parameter could also have been ckWARN to specify that the second
+ * category isn't on by default.
+ *
+ * This complicated expression removes complexity in each of the cases in the
+ * loop.  When called without a second category, it outputs a bunch of zeroes
+ * that the compiler should fold to nothing */
+#define NEED_MESSAGE(warning, extra_ckWARN, extra_category)             \
+            ((flags & UTF8_CHECK_ONLY)                                  \
+             ? 0                                                        \
+             : ((msgs)                                                  \
+                   /* Here are to return a warning message.  Choose the \
+                    * highest priority enabled category, or 'warning'   \
+                    * if neither is enabled */                          \
+                ? ((ckWARN_d(warning))                                  \
+                   ? warning                                            \
+                   : ((extra_ckWARN(extra_category)                     \
+                      ? extra_category                                  \
+                      : warning)))                                      \
+                   /* Here are to raise a warning if either category is \
+                    * enabled.  Return the highest priority enabled     \
+                    * one, or 0 if neither is enabled */                \
+                : ((ckWARN_d(warning)                                   \
+                   ? warning                                            \
+                   : ((extra_ckWARN(extra_category)                     \
+                      ? extra_category                                  \
+                      : 0))))))
```

Comment:
Wouldn't this be simpler if you invert the order of checks (i.e. check `msgs` on the inside)?

```c
             : ((ckWARN_d(warning))                                   \
                 ? warning                                            \
                 : ((extra_ckWARN(extra_category)                     \
                    ? extra_category                                  \
                    : (msgs) ? warning : 0))))                        \

```